mono/io-layer/processes.c

```diff
@@ -2754,7 +2754,32 @@ process_wait (gpointer handle, guint32 timeout, gboolean alertable)
 	/* We don't need to lock mono_processes here, the entry
 	 * has a handle_count > 0 which means it will not be freed. */
 	mp = process_handle->mono_process;
-	g_assert (mp);
+	if (!mp) {
+		guint32 start, now;
+
+		/* We've already done this one */
+		MONO_TRACE (G_LOG_LEVEL_DEBUG, MONO_TRACE_IO_LAYER, "%s (%p, %u): waiting on non-child process", __func__, handle, timeout);
+
+		for (start = now = mono_msec_ticks (); timeout == INFINITE || now - start < timeout || now == start; now = mono_msec_ticks ()) {
+			int status;
+			pid_t res = waitpid (pid, &status, WNOHANG);
```

Comment:
That brings the question of how much value bringing Process has. Since it's 100% implemented on top of the windows behavior.

I'm looking at the code for HasExited and it polls for the exit status by using a zero timeout.

I think we can support just that and avoid polling on ppl's back.